contracts/colonyNetwork/ColonyNetworkExtensions.sol

```diff
@@ -0,0 +1,135 @@
+    require(version == 1 || resolvers[_extensionId][version - 1] != address(0x0), "extension-manager-bad-version");
+
+    resolvers[_extensionId][version] = _resolver;
+    if (version == 1) { roles[_extensionId] = _roles; }
+
+    emit ExtensionAdded(_extensionId, version);
+  }
+
+  function installExtension(bytes32 _extensionId, address _colony, uint256 _version)
+    public
+    stoppable
+  {
+    require(resolvers[_extensionId][_version] != address(0x0), "extension-manager-bad-version");
+    require(installations[_extensionId][_colony] == address(0x0), "extension-manager-already-installed");
+    require(root(_colony), "extension-manager-unauthorized");
+
+    EtherRouter extension = new EtherRouter();
+    installations[_extensionId][_colony] = address(extension);
+
+    extension.setResolver(resolvers[_extensionId][_version]);
+    ColonyExtension(address(extension)).install(_colony);
+
+    emit ExtensionInstalled(_extensionId, _colony, _version);
+  }
+
+  function upgradeExtension(bytes32 _extensionId, address _colony, uint256 _newVersion)
+    public
+    stoppable
+  {
+    require(root(_colony), "extension-manager-unauthorized");
+    require(installations[_extensionId][_colony] != address(0x0), "extension-manager-not-installed");
+
+    address payable extension = installations[_extensionId][_colony];
+    require(_newVersion == ColonyExtension(extension).version() + 1, "extension-manager-bad-increment");
+    require(resolvers[_extensionId][_newVersion] != address(0x0), "extension-manager-bad-version");
+
+    EtherRouter(extension).setResolver(resolvers[_extensionId][_newVersion]);
+    ColonyExtension(extension).finishUpgrade();
+    assert(ColonyExtension(extension).version() == _newVersion);
+
+    emit ExtensionUpgraded(_extensionId, _colony, _newVersion);
+  }
+
+  function uninstallExtension(bytes32 _extensionId, address _colony)
+    public
+    stoppable
+  {
+    require(root(_colony), "extension-manager-unauthorized");
+    require(installations[_extensionId][_colony] != address(0x0), "extension-manager-not-installed");
+
+    ColonyExtension extension = ColonyExtension(installations[_extensionId][_colony]);
+    installations[_extensionId][_colony] = address(0x0);
+    extension.uninstall();
+
+    emit ExtensionUninstalled(_extensionId, _colony);
+  }
+
+  // Public view functions
+
+  function getExtensionRoles(bytes32 _extensionId)
+    public
+    view
+    returns (bytes32)
+  {
+    return roles[_extensionId];
+  }
+
+  function getExtensionResolver(bytes32 _extensionId, uint256 _version)
+    public
+    view
+    returns (address)
+  {
+    return resolvers[_extensionId][_version];
+  }
+
+  function getExtensionInstallation(bytes32 _extensionId, address _colony)
+    public
+    view
+    returns (address)
+  {
+    return installations[_extensionId][_colony];
+  }
+
+  // Internal functions
+
+  function root(address _colony) internal view returns (bool) {
```

Comment:
Sure! I'll also make it a modifer.

Reply:
Although... it would be more consistent with our conventions to add these functions to the IColony interface, which does the auth, and then have the colony make the call out to ColonyNetwork. What do you think of that?

Reply:
Do we do that for things in `IColony`? I know we do in `IMetaColony`, but that's special :smile:

Reply:
For example here:

```
  function registerColonyLabel(string memory colonyName, string memory orbitdb) public stoppable auth {
    IColonyNetwork(colonyNetworkAddress).registerColonyLabel(colonyName, orbitdb);
  }
```

Reply:
I say let's be consistent.